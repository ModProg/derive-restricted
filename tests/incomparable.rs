use derive_where::derive_where;

macro_rules! incomparable {
	($expr:expr) => {
		assert_ne!($expr, $expr);
		assert_eq!($expr.partial_cmp(&$expr), None)
	};
	($a:expr, $b:expr) => {
		assert_ne!($a, $b);
		assert_eq!($a.partial_cmp(&$b), None)
	};
}

#[test]
fn struct_() {
	#[derive(Debug)]
	#[derive_where(PartialEq, PartialOrd)]
	#[derive_where(incomparable)]
	struct Unit;
	incomparable!(Unit);

	#[derive(Debug)]
	#[derive_where(PartialEq, PartialOrd)]
	#[derive_where(incomparable)]
	struct Tuple();
	incomparable!(Tuple());

	#[derive(Debug)]
	#[derive_where(PartialEq, PartialOrd)]
	#[derive_where(incomparable)]
	struct Struct {}
	incomparable!(Struct {});
}

#[test]
fn enum_() {
	#[derive(Debug)]
	#[derive_where(PartialEq, PartialOrd)]
	#[derive_where(incomparable)]
	// Note that it is not possible to test an empty enum
	enum Empty {}

	#[derive(Debug)]
	#[derive_where(PartialEq, PartialOrd)]
	#[derive_where(incomparable)]
	enum Item {
		A,
		B,
	}
	incomparable!(Item::A);
	incomparable!(Item::A, Item::B);

	#[derive(Debug)]
	#[derive_where(PartialEq, PartialOrd)]
	enum Single {
		#[derive_where(incomparable)]
		A,
	}
	incomparable!(Single::A);

	#[derive(Debug)]
	#[derive_where(PartialEq, PartialOrd)]
	enum Enum {
		#[derive_where(incomparable)]
		IncomparableItem,
		#[derive_where(incomparable)]
		IncomparableTuple(u8),
		#[derive_where(incomparable)]
		IncomparableStruct {
			#[allow(unused)]
			a: u8,
		},
		ComparableItem,
		ComparableTuple(u8),
		ComparableStruct {
			a: u8,
		},
	}
	incomparable!(Enum::IncomparableItem);
	incomparable!(Enum::IncomparableItem, Enum::ComparableItem);
	incomparable!(Enum::IncomparableItem, Enum::ComparableTuple(1));
	incomparable!(Enum::IncomparableItem, Enum::ComparableStruct { a: 1 });

	incomparable!(Enum::IncomparableTuple(1));
	incomparable!(Enum::IncomparableTuple(1), Enum::IncomparableTuple(2));
	incomparable!(Enum::IncomparableTuple(1), Enum::ComparableItem);
	incomparable!(Enum::IncomparableTuple(1), Enum::ComparableTuple(1));
	incomparable!(Enum::IncomparableTuple(1), Enum::ComparableStruct { a: 1 });

	incomparable!(Enum::IncomparableStruct { a: 1 });
	incomparable!(
		Enum::IncomparableStruct { a: 1 },
		Enum::IncomparableStruct { a: 2 }
	);
	incomparable!(Enum::IncomparableStruct { a: 1 }, Enum::ComparableItem);
	incomparable!(Enum::IncomparableStruct { a: 1 }, Enum::ComparableTuple(1));
	incomparable!(
		Enum::IncomparableStruct { a: 1 },
		Enum::ComparableStruct { a: 1 }
	);

	assert_eq!(Enum::ComparableItem, Enum::ComparableItem);

	assert!(Enum::ComparableItem < Enum::ComparableTuple(0));
	assert_eq!(Enum::ComparableTuple(1), Enum::ComparableTuple(1));
	assert!(Enum::ComparableTuple(1) < Enum::ComparableTuple(2));

	assert!(Enum::ComparableTuple(1) < Enum::ComparableStruct { a: 0 });
	assert_eq!(
		Enum::ComparableStruct { a: 1 },
		Enum::ComparableStruct { a: 1 }
	);
	assert!(Enum::ComparableStruct { a: 1 } < Enum::ComparableStruct { a: 2 });
}
