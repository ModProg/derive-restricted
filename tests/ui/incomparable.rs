use derive_where::derive_where;

#[derive_where(PartialEq)]
#[derive_where(incomparable)]
enum IncomparableOnVariantAndEnum {
	#[derive_where(incomparable)]
	Variant,
}

#[derive_where(Debug)]
#[derive_where(incomparable)]
enum IncomparableWithoutPartialCmpTrait {
	Variant,
}

#[derive_where(Eq)]
#[derive_where(incomparable)]
enum IncomparableWithCmpTrait {
	Variant,
}

fn main() {}
