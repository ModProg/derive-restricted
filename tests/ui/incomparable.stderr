error: `incomparable` cannot be specified on both item and variant
 --> tests/ui/incomparable.rs:4:16
  |
4 |   #[derive_where(incomparable)]
  |  ________________^
5 | | enum IncomparableOnVariantAndEnum {
6 | |     #[derive_where(incomparable)]
  | |_______________________________^

error: `incomparable` is only supported if `PartialEq` or `PartialOrd` is being implemented
  --> tests/ui/incomparable.rs:11:16
   |
11 | #[derive_where(incomparable)]
   |                ^^^^^^^^^^^^

error: `incomparable` is not supported if `Eq` or `Ord` is being implemented
  --> tests/ui/incomparable.rs:17:16
   |
17 | #[derive_where(incomparable)]
   |                ^^^^^^^^^^^^
